Bukkit/src/main/java/com/plotsquared/bukkit/util/BukkitRegionManager.java

```diff
@@ -88,73 +84,29 @@ public boolean handleClear(
 
     @Override
     public int[] countEntities(@NonNull Plot plot) {
+        int[] count = new int[6];
         int[] existing = (int[]) plot.getMeta("EntityCount");
         if (existing != null && (System.currentTimeMillis() - (long) plot.getMeta("EntityCountTime") < 1000)) {
             return existing;
         }
         PlotArea area = plot.getArea();
         World world = BukkitUtil.getWorld(area.getWorldName());
-        Location bot = plot.getBottomAbs();
-        Location top = plot.getTopAbs();
-        int bx = bot.getX() >> 4;
-        int bz = bot.getZ() >> 4;
-
-        int tx = top.getX() >> 4;
-        int tz = top.getZ() >> 4;
-
-        int size = tx - bx << 4;
-
-        Set<Chunk> chunks = new HashSet<>();
-        for (int X = bx; X <= tx; X++) {
-            for (int Z = bz; Z <= tz; Z++) {
-                if (world.isChunkLoaded(X, Z)) {
-                    chunks.add(world.getChunkAt(X, Z));
-                }
-            }
-        }
-
-        boolean doWhole = false;
-        List<Entity> entities = null;
-        if (size > 200 && chunks.size() > 200) {
-            entities = world.getEntities();
-            if (entities.size() < 16 + size / 8) {
-                doWhole = true;
-            }
+        if(world == null) {
+            return count;
         }
 
-        int[] count = new int[6];
-        if (doWhole) {
-            for (Entity entity : entities) {
-                org.bukkit.Location location = entity.getLocation();
-                PaperLib.getChunkAtAsync(location).thenAccept(chunk -> {
-                    if (chunks.contains(chunk)) {
-                        int X = chunk.getX();
-                        int Z = chunk.getZ();
-                        if (X > bx && X < tx && Z > bz && Z < tz) {
-                            count(count, entity);
-                        } else {
-                            Plot other = area.getPlot(BukkitUtil.adapt(location));
-                            if (plot.equals(other)) {
-                                count(count, entity);
+        for (final CuboidRegion region : plot.getRegions()) {
+            for (int x = region.getMinimumPoint().getX() >> 4; x <= region.getMaximumPoint().getX() >> 4; x++) {
+                for (int z = region.getMinimumPoint().getZ() >> 4; z <= region.getMaximumPoint().getZ() >> 4; z++) {
+                    world.getChunkAtAsync(x, z).thenAccept(chunk -> {
+                        final Entity[] entities = chunk.getEntities();
```

Comment:
Not all entities in the chunk also have to be in the plot. This needs to be filtered again.